src/gpuarray_reduction.c

```diff
@@ -0,0 +1,548 @@
+		GA_POINTER, /* srcSteps */
+		GA_POINTER, /* srcSize */
+		GA_POINTER, /* numBlk */
+		GA_POINTER, /* dstMax */
+		GA_POINTER, /* dstMaxSteps */
+		GA_POINTER, /* dstArgmax */
+		GA_POINTER  /* dstArgmaxSteps */
+	};
+	const size_t l = strlen(s);
+	GpuKernel kernel;
+	GpuKernel_init(&kernel, 0, 1, &s, &l, "maxandargmax",
+	               8, ARG_TYPECODE, 0, (char**)0);
+	
+	/* Invoke it. */
+	invokeMaxAndArgmax(&kernel, src, isReduced);
+
+	/* Return error code */
+	return GA_NO_ERROR;
+}
+
+/**
+ * Count the number of dimensions to be reduced.
+ */
+
+static int  getRdxIdx(const int numIdx, const int* isReduced){
+	int i, countReduced;
+	for(i=0, countReduced = 0;i<numIdx;i++){
+		countReduced += !!isReduced[i];
+	}
+	return countReduced;
+}
+
+/**
+ * @brief Generate the kernel code for MaxAndArgmax.
+ * 
+ * @param [in]  numIdx
+ * @param [in]  isReduced
+ * @param [in]  dstMaxType
+ * @param [in]  dstArgmaxType
+ * @return A free()'able string containing source code implementing the
+ *         kernel, or else NULL.
+ */
+
+static char* genkernel_maxandargmax(const int       numIdx,
+                                    const int*      isReduced,
+                                    const char*     dstMaxType,
+                                    const char*     dstArgmaxType){
+	/* Obtain the parameters of the reduction. */
+	GEN_KERNEL_CTX ctx;
+	ctx.numIdx        = numIdx;
+	ctx.isReduced     = isReduced;
+	ctx.numRedIdx     = getRdxIdx(ctx.numIdx, ctx.isReduced);
+	ctx.numFreeIdx    = ctx.numIdx - ctx.numRedIdx;
+	ctx.dstMaxType    = dstMaxType;
+	ctx.dstArgmaxType = dstArgmaxType;
+	
+	/**
+	 * Allocate string buffer.
+	 */
+	
+	strb* s = strb_alloc(5*1024);
+	if(!s){return NULL;}
+	
+	/**
+	 * Generate kernel
+	 */
+	
+	appendKernel(s, &ctx);
+	
+	/**
+	 * Return the kernel.
+	 */
+	
+	char* kernelSource = strb_cstr(s);
+	strb_free(s);
+	return kernelSource;
+}
+
+static void  appendKernel           (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	appendTypedefs         (s, ctx);
+	appendPrototype        (s, ctx);
+	strb_appends           (s, "{\n");
+	appendIndexDeclarations(s, ctx);
+	appendRangeCalculations(s, ctx);
+	appendLoops            (s, ctx);
+	strb_appends           (s, "}\n");
+}
+static void  appendTypedefs         (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	strb_appends(s, "/* Typedefs */\n");
+	strb_appendf(s, "typedef %s     T;/* The type of the array being processed. */\n", ctx->dstMaxType);
+	strb_appendf(s, "typedef %s     X;/* Index type: signed 32/64-bit. */\n",          ctx->dstArgmaxType);
+	strb_appends(s, "\n");
+	strb_appends(s, "\n");
+	strb_appends(s, "\n");
+}
+static void  appendPrototype        (strb*           s,
+                                     GEN_KERNEL_CTX* ctx){
+	strb_appends(s, "__global__ void maxandargmax(const T*        src,\n");
+	strb_appends(s, "                             const X*        srcSteps,\n");
```

Comment:
@abergeron `X`, which is typedef'ed to `ctx->dstArgmaxType`, is currently always 64 bits due to the later's initialization as `gpuarray_get_type(GA_LONG)       -> cluda_name` (Will get changed to `GA_SIZE`). If in the future this will change, we can set that to 32 bits easily.